openshift-kube-apiserver/admission/storage/csiinlinevolumesecurity/admission_test.go

```diff
@@ -0,0 +1,747 @@
+			namespace:     testNamespaceNoLabels(),
+			driver:        testCSIDriverNoLabels(),
+			expectedError: fmt.Errorf(privNamespaceRequiredError),
+		},
+		{
+			name:      "should allow pods with no volumes",
+			obj:       testPod(),
+			namespace: testNamespaceNoLabels(),
+			driver:    testCSIDriverNoLabels(),
+		},
+		{
+			name:      "should allow pods with only persistent volume claims",
+			obj:       testPodWithPVC(),
+			namespace: testNamespaceNoLabels(),
+			driver:    testCSIDriverNoLabels(),
+		},
+		{
+			name:      "should allow pods with inline volumes when running in a privileged namespace",
+			obj:       testPodWithInlineVol(),
+			namespace: testNamespacePrivileged(),
+			driver:    testCSIDriverNoLabels(),
+		},
+		{
+			name:      "should allow pods with inline volumes in a restricted namespace when the driver uses the restricted label",
+			obj:       testPodWithInlineVol(),
+			namespace: testNamespaceRestricted(),
+			driver:    testCSIDriverRestricted(),
+		},
+		{
+			name:      "should allow pods with inline volumes in a baseline namespace when the driver uses the baseline label",
+			obj:       testPodWithInlineVol(),
+			namespace: testNamespaceBaseline(),
+			driver:    testCSIDriverBaseline(),
+		},
+		{
+			name:      "should allow pods with inline volumes in a baseline namespace when the driver uses the restricted label",
+			obj:       testPodWithInlineVol(),
+			namespace: testNamespaceBaseline(),
+			driver:    testCSIDriverRestricted(),
+		},
+		{
+			name:          "should deny pod admission with inline volumes in a baseline namespace when the driver uses the privileged label",
+			obj:           testPodWithInlineVol(),
+			namespace:     testNamespaceBaseline(),
+			driver:        testCSIDriverPrivileged(),
+			expectedError: fmt.Errorf(privNamespaceRequiredError),
+		},
+		{
+			name:          "should deny pod admission with inline volumes if the CSI driver is not found",
+			obj:           testPodWithInvalidDriverName(),
+			namespace:     testNamespaceRestricted(),
+			driver:        testCSIDriverRestricted(),
+			expectedError: fmt.Errorf("failed to get csi driver: csidriver.storage.k8s.io \"invalid-csi-driver\" not found"),
+		},
+		{
+			name:          "should deny pod admission if the CSI driver has an invalid profile label",
+			obj:           testPodWithInlineVol(),
+			namespace:     testNamespaceBaseline(),
+			driver:        testCSIDriverInvalid(),
+			expectedError: fmt.Errorf("invalid label security.openshift.io/csi-ephemeral-volume-profile for CSIDriver test-driver: must be one of privileged, baseline, restricted"),
+		},
+		{
+			name:          "should deny pod admission if the namespace has an invalid profile label",
+			obj:           testPodWithInlineVol(),
+			namespace:     testNamespaceInvalid(),
+			driver:        testCSIDriverRestricted(),
+			expectedError: fmt.Errorf("Invalid value: \"invalid-value\": must be one of privileged, baseline, restricted"),
+		},
+		/* PodTemplate */
```

Comment:
Since all the test variants have just different `createPodControllerObject` argument, would it be possible to generate the tests via a `for` loop over `coreapi.PodTemplate{}`, `coreapi.ReplicationController{}`, ..., `batchapi.CronJob{}`?

Reply:
Yes, that's much better, see [this commit](https://github.com/openshift/kubernetes/commit/f0501d4ac74983f23a9ccc9cb0a969b9231ca51a).